pkgs/profpatsch/nman/nman.rs

```diff
@@ -0,0 +1,318 @@
+    // This means we realise the least amount of outputs
+    // necessary
+    //
+    // TODO(sterni): change sorting depending on section:
+    //               "3" and "3p" should prioritize DevMan
+    drvs.sort_unstable_by(|a, b| a.output.cmp(&b.output));
+
+    if drvs.len() <= 0 {
+        return Err(NmanError::NixParseError);
+    }
+
+    for drv in drvs {
+        let man_file = build_man_page(drv, section, page, &tmpdir)?;
+
+        match man_file {
+            None => continue,
+            Some(f) => {
+                let res = Command::new("man")
+                                  .arg("-l").arg(f)
+                                  .spawn()
+                                  .and_then(|mut c| c.wait())
+                                  .map(|c| c.success());
+
+                return match res {
+                    Ok(true) => Ok(()),
+                    Ok(false) => Err(NmanError::Man),
+                    Err(_) => Err(NmanError::Execution),
+                };
+            },
+        }
+    }
+
+    Err(NmanError::NotFound)
+}
+
+fn parse_man_section(section: &str) -> Result<&str, &str> {
+    match section {
+        "3p" => Ok(section),
+        _ => match u8::from_str_radix(section, 10) {
+            Ok(_) => Ok(section),
+            Err(_)  => Err("Invalid man section: not a number and not \"3p\""),
+        },
+    }
+}
+
+enum CliAction<'a> {
+    Usage,
+    // attribute, section, page
+    // TODO(sterni): section should be an option type, so we can implement
+    //               the search logic as man(1) does. Also worth considering
+    //               would be to not find the man page ourselves, but to just
+    //               set MANPATH and let man(1) do the searching. Use case
+    //               would be to allow
+    //                  nman libunwind unw_apply_reg_state
+    //              to work like
+    //                  nman libunwind 3 unw_apply_reg_state
+    Man(&'a str, &'a str, &'a str),
+}
+
+fn main() -> std::io::Result<()> {
+    fn dispatch_action(progname: &str, action: CliAction) -> std::io::Result<()> {
+        match action {
+            CliAction::Usage => {
+                println!("Usage: {} ATTR [SECTION | PAGE] [PAGE]", progname);
+                Ok(())
+            },
+            CliAction::Man(attr, section, page) =>
+                match open_man_page(attr, section, page) {
+                    Ok(_) => Ok(()),
+                    // TODO(sterni): print error message
+                    Err(t) => {
+                        println!("error: {}", t.msg());
+                        std::process::exit(t.code())
+                    },
+                },
+        }
+    }
+
+    let (opts, args) : (Vec<String>, Vec<String>) =
+            std::env::args().partition(|s| s.starts_with("-"));
+
+    let mut action : Result<CliAction, &str> = match args.len() {
+        2 => Ok(CliAction::Man(&args[1], "1", &args[1])),
```

Comment:
I think the "1" should rather be a `Section::Unknown` here, so we can pass it to man in the way that it does its usual “open the first man page in order”.

Reply:
This also occurred to me and I've spelled my thoughts out a bit in the declaration of `CliAction`. I think we have two options for implementing this:

* Implement the man page search logic ourselves in `build_man_page`. This may be a bit annoying, but would also be a opportunity to implement the multiple extensions support more cleanly.
* Set `MANPATH` when invoking `man`. This would be very convenient and adhere to DRY as we can just reuse the existing implementation of man page searching. However we would need to restructure `build_man_page` as we wouldn't be able to return a file path in all cases and also we wouldn't be able to distinguish between “man page doesn't exist” and “man just failed for some reason” as `mandoc`'s `man(1)` and `man-db`'s `man(1)` don't use the same conventions for exit codes (although we could probably assume that exit code == 5 || == 10 means not found).

Reply:
Ah, the problem is to check whether a manpage exists. I think we should implement at least a bit of a search logic because we want to optionally build/download more outputs. So something like “we found multiple sections that could match, opening section N” to stderr.

Reply:
I'm also fairly convinced that we should implement the logic ourselves, but I think we can just stop searching as soon we find  a match since this is also the behavior of `man(1)`, e. g. if I run `nman lowdown` it is not really surprising that it opens `lowdown(1)` instead of `lowdown(3)`. The most important point here is to not fail if a match exists. If an user gets a surprising match, they will know how to deal with it since it is basically part of the `man(1)` UX at this point.

Reply:
Done in 6c7d80af4e092f31eb6cfcaadd82c90bd36faaef!